# Satisfaction, Loyalty and its drivers by client segment

## Background

Customer satisfaction plays an important role in building strong client relationships, encouraging loyalty and supporting long-term business growth. Understanding how clients view the company’s services can help identify areas of strength and opportunities for improvement.

This analysis examines customer satisfaction across different client segments and years. It considers the quality of the company’s services as well as clients’ willingness to recommend the company.

## Analysis Overview

The analysis focuses on three client segments:

* **Government organisations (Govt)**
* **Non-Profit Organisations (NPO)**
* **Private organisations**

Customer satisfaction is evaluated using four service dimensions:

* Presales & Partnership
* Technical Expertise
* Project Delivery
* Post-Sales Support

Each service dimension is rated from **1 to 5**, where a higher value represents greater customer satisfaction.

The analysis also uses **NPS Rating**, which is recorded from **0 to 10**. NPS represents a client’s willingness to recommend the company and is used as an indication of customer advocacy, loyalty and overall experience. A higher NPS Rating generally suggests a more positive client relationship.

As the dataset contains individual NPS ratings, this analysis uses the **average NPS Rating** instead of calculating the formal Net Promoter Score.

## Purpose of the Analysis

The purpose of this analysis is to:

1. Compare customer satisfaction across different client segments.
2. Examine how satisfaction and NPS ratings change over time.
3. Identify client segments with relatively high or low satisfaction.
4. Determine which service dimensions are most strongly associated with NPS Rating.
5. Highlight areas where service improvements may strengthen customer satisfaction and loyalty.

## Connection to the Main Problem

The team’s main objective is to determine which client segments generate the highest gross profit while maintaining strong customer satisfaction.

This analysis provides the **customer satisfaction component** of the main problem. Its findings can be combined with the profitability analysis to identify client segments that provide both strong financial value and positive customer outcomes.



## Data preparation and validation

The analysis uses the cleaned, merged client dataset. Each row represents a **client-year observation**, so the same client may legitimately appear in different years.

- Service dimensions are rated on a **1–5 scale**.
- NPS Rating is recorded on a **0–10 scale**.
- Client Type is the main segmentation variable used in these visualisations.

The checks below confirm the dataset shape, inspect the available fields and verify that there is no more than one record for the same Client ID and Year combination.

In [10]:
import pandas as pd
import plotly.express as px

df = pd.read_excel("merged_cleaned.xlsx")

print("Shape:", df.shape)
print(
    "Duplicate client-year records:",
    df.duplicated(subset=["CLIENT ID", "YEAR"]).sum()
)

display(df.head())

Shape: (392, 16)
Duplicate client-year records: 0


,CLIENT ID,TYPE,COMMENCEMENT DATE,STAFF STRENGTH,SECTOR,COUNTRY,YEAR,PRESALES AND PARTNERSHIP,TECHNICAL EXPERTISE,PROJECT DELIVERY,POST-SALES SUPPORT,NPS RATING,NPS RATING.1,NPS RATING.2,SOFTWARE,MANPOWER
0,G1-001,Govt,2020-01-19 00:00:00,> 200,Finance,Singapore,2021,3,4,4,5,8,148180,17321,25562,41201
1,G1-001,Govt,2020-01-19 00:00:00,> 200,Finance,Singapore,2022,5,5,3,4,9,197313,23604,32875,50085
2,G1-001,Govt,2020-01-19 00:00:00,> 200,Finance,Singapore,2023,5,4,3,4,7,349899,51332,71695,89089
3,G1-001,Govt,2020-01-19 00:00:00,> 200,Finance,Singapore,2024,5,5,3,4,8,268821,40969,56452,76545
4,G1-001,Govt,2020-01-19 00:00:00,> 200,Finance,Singapore,2025,5,4,5,5,10,138115,14516,20889,28968


In [11]:
print(
    "Duplicate client-year records:",
    df.duplicated(subset=["CLIENT ID", "YEAR"]).sum()
)

Duplicate client-year records: 0


### Validation result

The cleaned dataset contains **392 client-year observations and 16 variables**, with **0 duplicate Client ID–Year records**. This supports fair yearly comparisons without double-counting the same client within a year.

The next two cells define the four service dimensions and calculate an initial client-type summary before the interactive visualisations are created.


In [12]:
service_columns = [
    "PRESALES AND PARTNERSHIP",
    "TECHNICAL EXPERTISE",
    "PROJECT DELIVERY",
    "POST-SALES SUPPORT"
]

In [13]:
type_summary = (
    df.groupby("TYPE")[service_columns]
      .mean()
      .round(2)
      .reset_index()
)

display(type_summary)

,TYPE,PRESALES AND PARTNERSHIP,TECHNICAL EXPERTISE,PROJECT DELIVERY,POST-SALES SUPPORT
0,Govt,4.09,3.94,4.35,3.91
1,NPO,3.76,3.88,4.26,4.12
2,Private,3.96,3.92,4.01,3.80


## Visualisation 1 — Service Ratings by Client Type and Period

### Purpose

This visualisation compares the average service ratings of Government, NPO and Private clients from 2021 to 2025.

The service ratings represent clients’ satisfaction with four areas of the company’s service:

* **Presales & Partnership:** The quality of engagement before a project begins and the company’s ability to develop a strong working relationship with the client.
* **Technical Expertise:** The client’s perception of the company’s technical knowledge, professional skills and ability to provide suitable solutions.
* **Project Delivery:** How effectively the company plans, manages and completes its projects.
* **Post-Sales Support:** The quality of assistance provided after project completion, including ongoing support and issue resolution.

Each dimension is rated from **1 to 5**, where a higher rating represents greater satisfaction. The bars show the **average rating** given by clients within each client type and period.

### What the Visualisation Shows

The grouped bars allow the four service dimensions to be compared within each client type. This helps identify:

* the strongest and weakest service areas for each client segment;
* whether different client types have different service experiences;
* whether satisfaction improves or declines over time; and
* which service areas may require greater attention.

The **Overall** view summarises all years, while the yearly views show how the ratings change between 2021 and 2025.

### How to Interact

* Press **Play** to animate the changes from 2021 to 2025 and the Overall view.
* Drag the **period slider** to select a specific year.
* Hover over a bar to view the client type, service dimension, average rating and number of records.
* Click a service dimension in the legend to show or hide it.

The y-axis remains fixed from **0 to 5** so that ratings can be compared fairly across different years and client types.

### Interpretation

A taller bar represents a higher average level of satisfaction with that service dimension. Large differences between the bars may indicate that clients are more satisfied with some areas of service than others.

These comparisons help the company recognise service strengths that should be maintained and weaknesses that could be improved to support stronger customer satisfaction and loyalty.



In [14]:
import pandas as pd
import plotly.express as px


# -------------------------------------------------
# 1. Settings
# -------------------------------------------------

service_columns = [
    "PRESALES AND PARTNERSHIP",
    "TECHNICAL EXPERTISE",
    "PROJECT DELIVERY",
    "POST-SALES SUPPORT"
]

service_label_map = {
    "PRESALES AND PARTNERSHIP": "Presales & Partnership",
    "TECHNICAL EXPERTISE": "Technical Expertise",
    "PROJECT DELIVERY": "Project Delivery",
    "POST-SALES SUPPORT": "Post-Sales Support"
}

service_order = list(service_label_map.values())
type_order = ["Govt", "NPO", "Private"]

service_colors = {
    "Presales & Partnership": "#636EFA",
    "Technical Expertise": "#EF553B",
    "Project Delivery": "#00CC96",
    "Post-Sales Support": "#AB63FA"
}


# -------------------------------------------------
# 2. Calculate overall averages
# -------------------------------------------------

overall_summary = (
    df.groupby("TYPE", as_index=False)[service_columns]
      .mean()
      .round(2)
)

overall_counts = (
    df.groupby("TYPE")
      .size()
      .rename("Records")
      .reset_index()
)

overall_summary = overall_summary.merge(
    overall_counts,
    on="TYPE"
)

overall_long = overall_summary.melt(
    id_vars=["TYPE", "Records"],
    value_vars=service_columns,
    var_name="Service Dimension",
    value_name="Average Rating"
)

overall_long["Service Dimension"] = (
    overall_long["Service Dimension"]
    .replace(service_label_map)
)

overall_long["Period"] = "Overall"


# -------------------------------------------------
# 3. Calculate yearly averages
# -------------------------------------------------

yearly_summary = (
    df.groupby(["YEAR", "TYPE"], as_index=False)[service_columns]
      .mean()
      .round(2)
)

yearly_counts = (
    df.groupby(["YEAR", "TYPE"])
      .size()
      .rename("Records")
      .reset_index()
)

yearly_summary = yearly_summary.merge(
    yearly_counts,
    on=["YEAR", "TYPE"]
)

yearly_long = yearly_summary.melt(
    id_vars=["YEAR", "TYPE", "Records"],
    value_vars=service_columns,
    var_name="Service Dimension",
    value_name="Average Rating"
)

yearly_long["Service Dimension"] = (
    yearly_long["Service Dimension"]
    .replace(service_label_map)
)

yearly_long["Period"] = (
    yearly_long["YEAR"]
    .astype(int)
    .astype(str)
)


# -------------------------------------------------
# 4. Combine yearly and overall results
# -------------------------------------------------

animation_data = pd.concat(
    [
        yearly_long.drop(columns="YEAR"),
        overall_long
    ],
    ignore_index=True
)

animation_data["Animation Group"] = (
    animation_data["TYPE"]
    + " — "
    + animation_data["Service Dimension"]
)

period_order = [
    str(int(year))
    for year in sorted(df["YEAR"].unique())
] + ["Overall"]


# -------------------------------------------------
# 5. Create animated chart
# -------------------------------------------------

fig = px.bar(
    animation_data,
    x="TYPE",
    y="Average Rating",
    color="Service Dimension",
    barmode="group",

    animation_frame="Period",
    animation_group="Animation Group",

    text="Average Rating",
    custom_data=["Records"],

    category_orders={
        "Period": period_order,
        "TYPE": type_order,
        "Service Dimension": service_order
    },

    color_discrete_map=service_colors,

    labels={
        "TYPE": "Client Type",
        "Average Rating": "Average Rating"
    },

    range_y=[0, 5.4],

    title="Service Ratings by Client Type: 2021–2025 and Overall"
)


# -------------------------------------------------
# 6. Hover and data-label formatting
# -------------------------------------------------

trace_format = dict(
    texttemplate="%{y:.2f}",
    textposition="outside",
    cliponaxis=False,

    hovertemplate=(
        "<b>Client Type: %{x}</b><br>"
        "Service: %{fullData.name}<br>"
        "Average Rating: %{y:.2f} / 5<br>"
        "Records Used: %{customdata[0]}"
        "<extra></extra>"
    )
)

fig.update_traces(**trace_format)

# Apply the same formatting to every animation frame
for frame in fig.frames:
    for trace in frame.data:
        trace.update(**trace_format)


# -------------------------------------------------
# 7. Adjust animation speed
# -------------------------------------------------

play_settings = (
    fig.layout.updatemenus[0]
       .buttons[0]
       .args[1]
)

play_settings["frame"] = {
    "duration": 1400,
    "redraw": True
}

play_settings["transition"] = {
    "duration": 700,
    "easing": "cubic-in-out"
}

play_settings["fromcurrent"] = True

fig.layout.sliders[0].currentvalue.prefix = "Period: "
fig.layout.sliders[0].pad.t = 50


# -------------------------------------------------
# 8. Final formatting
# -------------------------------------------------

fig.update_layout(
    template="plotly_white",

    title=dict(
        x=0.5,
        xanchor="center"
    ),

    xaxis_title="Client Type",
    yaxis_title="Average Rating",

    legend_title="Service Dimension",

    transition=dict(
        duration=700,
        easing="cubic-in-out"
    ),

    margin=dict(
        t=100,
        b=130
    )
)

fig.update_yaxes(
    range=[0, 5.4],
    dtick=1
)


# -------------------------------------------------
# 9. Display graph
# -------------------------------------------------

fig.show(
    config={
        "displayModeBar": True,
        "displaylogo": False,
        "scrollZoom": True,
        "responsive": True,

        "toImageButtonOptions": {
            "format": "png",
            "filename": "service_ratings_story",
            "scale": 2
        }
    }
)

### Key Findings from Visualisation 1

#### Overall Performance

Average service ratings range from **3.76 to 4.35 out of 5**, indicating generally positive satisfaction across all client types.

**Project Delivery is the strongest overall dimension** for every segment:

* Government: **4.35**
* NPO: **4.26**
* Private: **4.01**

It also reaches its highest rating for all segments in **2025**: Government 4.75, NPO 4.54 and Private 4.18. This suggests that Project Delivery has strengthened over time and should remain a key service priority.

#### Trends by Client Type (2021 to 2025)

| Client Type    | Main Trend                                                                                                                                                   |
| -------------- | ------------------------------------------------------------------------------------------------------------------------------------------------------------ |
| **Government** | Project Delivery improves strongly from 3.67 in 2021 to 4.75 in 2025. Presales also improves overall, while Post-Sales Support remains relatively unchanged. |
| **NPO**        | Project Delivery increases from 4.00 to 4.54. Post-Sales Support improves after 2021, but Presales & Partnership falls from 4.50 to 3.85.                    |
| **Private**    | Ratings are more stable. Technical Expertise and Project Delivery improve, while Post-Sales Support remains the weakest dimension at 3.83 in 2025.           |

Overall, **Project Delivery shows the clearest positive trend** across all three client types. Technical Expertise also improves among Government and Private clients, while Presales and Post-Sales Support show less consistent changes.

#### Business Implications

The company should:

1. Maintain its strong Project Delivery performance.
2. Review the practices used in 2025, when Project Delivery reached its highest ratings.
3. Strengthen Presales & Partnership for NPO clients.
4. Improve Post-Sales Support for Private clients.
5. Apply different improvement strategies to each client type instead of using one general approach.

#### Limitations

The chart displays average ratings, which may hide differences between individual clients. The client segments also contain different numbers of records, meaning that changes in smaller groups may be influenced by only a few responses.

The trends show how ratings changed over time but do not explain what caused those changes.



## Visualisation 2 — Average NPS Rating by Client Type and Year

### Purpose

This heatmap compares the average NPS Rating of Government, NPO and Private clients from 2021 to 2025, together with an Overall average.

NPS Rating represents a client’s willingness to recommend the company. It is used as an indicator of customer advocacy, loyalty and overall experience. A higher average rating suggests that clients are more likely to view and recommend the company positively.

### What the Visualisation Shows

Each heatmap cell represents the average NPS Rating for one client type and period. The visualisation helps identify:

* which client type has the highest or lowest average NPS Rating;
* whether customer advocacy improves or declines over time;
* years with unusually high or low ratings; and
* client segments that may require additional attention.

The **Overall** column summarises the average across all years, while the yearly columns show how NPS Rating changes from 2021 to 2025.

### How to Interpret the Heatmap

The colour of each cell represents the average NPS Rating:

* Higher values indicate stronger customer advocacy and loyalty.
* Lower values may indicate weaker advocacy or areas requiring improvement.
* The printed value allows the exact average to be compared across cells.

The colour scale focuses on the range of ratings found in the dataset. Therefore, the colour differences may appear larger than the actual numerical differences, and the printed values should always be considered.

### How to Interact

Hover over a cell to view:

* the client type;
* the selected year or Overall period;
* the average NPS Rating; and
* the number of client-year records used to calculate the average.

The record count is important because averages based on fewer responses may be more affected by individual ratings.

> **Measurement note:** This chart uses the average individual **NPS Rating** on a 0–10 scale. It does not show the formal Net Promoter Score, which is calculated as the percentage of Promoters minus the percentage of Detractors.


In [15]:
import pandas as pd
import plotly.graph_objects as go


# -------------------------------------------------
# 1. Set display order
# -------------------------------------------------

type_order = [
    client_type
    for client_type in ["Govt", "NPO", "Private"]
    if client_type in df["TYPE"].unique()
]

years = sorted(
    df["YEAR"].dropna().astype(int).unique()
)


# -------------------------------------------------
# 2. Calculate average NPS rating
# -------------------------------------------------

nps_heatmap = (
    df.groupby(["TYPE", "YEAR"])["NPS RATING"]
      .mean()
      .round(2)
      .unstack("YEAR")
      .reindex(type_order)
)

nps_heatmap.columns = [
    str(int(year))
    for year in nps_heatmap.columns
]

nps_heatmap["Overall"] = (
    df.groupby("TYPE")["NPS RATING"]
      .mean()
      .round(2)
      .reindex(type_order)
)


# -------------------------------------------------
# 3. Calculate number of records for hover
# -------------------------------------------------

record_heatmap = (
    df.groupby(["TYPE", "YEAR"])
      .size()
      .unstack("YEAR")
      .reindex(type_order)
)

record_heatmap.columns = [
    str(int(year))
    for year in record_heatmap.columns
]

record_heatmap["Overall"] = (
    df.groupby("TYPE")
      .size()
      .reindex(type_order)
)


# -------------------------------------------------
# 4. Create Graph Objects heatmap
# -------------------------------------------------

fig2 = go.Figure(
    data=go.Heatmap(
        x=nps_heatmap.columns,
        y=nps_heatmap.index,
        z=nps_heatmap.values,

        text=nps_heatmap.values,
        texttemplate="%{text:.2f}",
        textfont=dict(
            size=16,
            color="#17324D"
        ),

        customdata=record_heatmap.values,

        zmin=7,
        zmax=9,

        colorscale=[
            [0.00, "#E76F51"],
            [0.25, "#F4A261"],
            [0.50, "#F6D365"],
            [0.75, "#82C9A5"],
            [1.00, "#2A9D8F"]
        ],

        colorbar=dict(
            title="Average NPS<br>Rating",
            tickvals=[7, 7.5, 8, 8.5, 9],
            ticktext=["7.0", "7.5", "8.0", "8.5", "9.0+"],
            thickness=18,
            len=0.85
        ),

        xgap=5,
        ygap=5,

        hovertemplate=(
            "<b>Client Type: %{y}</b><br>"
            "Period: %{x}<br>"
            "Average NPS Rating: %{z:.2f} / 10<br>"
            "Records Used: %{customdata}"
            "<extra></extra>"
        )
    )
)

# -------------------------------------------------
# 5. Format graph
# -------------------------------------------------

fig2.update_layout(
    title=dict(
        text="Average NPS Rating by Client Type and Year",
        x=0.5,
        xanchor="center"
    ),

    xaxis=dict(
        title="Period",
        side="bottom",
        type="category"
    ),

    yaxis=dict(
        title="Client Type",
        autorange="reversed",
        type="category"
    ),

    template="plotly_white",

    hoverlabel=dict(
        bgcolor="white",
        font_size=13
    ),

    margin=dict(
        l=100,
        r=120,
        t=90,
        b=80
    ),

    height=430
)


# -------------------------------------------------
# 6. Display graph
# -------------------------------------------------

fig2.show(
    config={
        "displayModeBar": True,
        "displaylogo": False,
        "responsive": True,

        "toImageButtonOptions": {
            "format": "png",
            "filename": "nps_rating_heatmap",
            "scale": 2
        }
    }
)

### Key Findings from Visualisation 2

#### Overall NPS Performance

NPO clients have the highest overall average NPS Rating at **8.12**, followed very closely by Government clients at **8.11**. The difference of only 0.01 means that their overall results are practically similar.

Private clients record a lower overall average of **7.81**. Although this is still a generally positive rating, it suggests that Private clients may be less willing to recommend the company compared with Government and NPO clients.

#### Trends by Client Type

**Government clients**

The average NPS Rating decreases slightly from **7.67 in 2021** to **7.56 in 2023**. It then improves to **8.00 in 2024** and rises sharply to **8.65 in 2025**.

The 2025 result is the highest individual value in the heatmap, suggesting a strong recent improvement in customer advocacy among Government clients.

**NPO clients**

NPO ratings fluctuate more across the period. The rating rises from **7.50 in 2021** to **8.33 in 2022**, falls to **7.83 in 2023** and **7.80 in 2024**, before reaching its highest value of **8.54 in 2025**.

Although the yearly results are inconsistent, NPO clients achieve the highest overall average NPS Rating.

**Private clients**

Private clients show the most consistent improvement. Their average NPS Rating increases every year from **7.38 in 2021** to **8.02 in 2025**.

Despite having the lowest overall average, the continuous improvement suggests that Private clients’ willingness to recommend the company is gradually strengthening.

#### Overall Trend

The year **2025 is the strongest period for all three client types**:

* Government: **8.65**
* NPO: **8.54**
* Private: **8.02**

This shared improvement may indicate that the overall client experience became more positive in 2025. The company should investigate whether any service improvements, project practices or relationship-management strategies were introduced during this period.

However, the chart alone cannot determine what caused the increase.

#### Business Implications

The company should:

1. Identify and maintain the practices that contributed to the strong 2025 results.
2. Continue improving the Private client experience, as this segment has the lowest overall NPS Rating.
3. Maintain its strong relationships with Government and NPO clients.
4. Examine why NPO ratings fluctuate more between years to determine whether their experience is inconsistent.
5. Convert satisfied clients into stronger advocates, as the overall averages remain below the Promoter threshold of 9.

#### Limitations

The chart displays **average NPS Ratings**, not the formal Net Promoter Score. An average rating cannot show the actual percentages of Promoters, Passives and Detractors within each segment.

The client groups may also contain different numbers of records, meaning that averages from smaller groups may be more affected by individual responses.

Finally, the colour scale focuses on ratings from approximately 7 to 9. This makes differences easier to see but can make small numerical gaps appear larger than they are. The printed values should therefore be considered alongside the colours.



## Visualisation 3 — Service Dimensions Associated with NPS Rating

### Purpose

This visualisation examines how strongly each service dimension is associated with NPS Rating. It helps identify which aspects of the client experience may be most relevant to customer advocacy and loyalty.

The four service dimensions examined are:

* Presales & Partnership
* Technical Expertise
* Project Delivery
* Post-Sales Support

The analysis is shown for all clients together and separately for Government, NPO and Private clients. Comparing the segments helps determine whether different client types have different satisfaction drivers.

### Correlation Method

The chart uses **Spearman correlation** to measure the direction and strength of the relationship between each service rating and NPS Rating.

Spearman correlation is suitable because the service dimensions use an ordered 1–5 rating scale. It examines whether higher service ratings generally occur together with higher NPS ratings without requiring the relationship to be perfectly linear.

The correlation value ranges from **−1 to +1**:

* A value closer to **+1** indicates a stronger positive relationship. Higher service ratings generally occur with higher NPS ratings.
* A value close to **0** indicates little or no consistent relationship.
* A value closer to **−1** indicates a stronger negative relationship. Higher service ratings generally occur with lower NPS ratings.

A longer positive bar therefore represents a service dimension that has a stronger positive association with customer advocacy and loyalty.

### What the Visualisation Shows

The chart helps identify:

* which service dimension has the strongest overall association with NPS Rating;
* whether the strongest loyalty driver differs across client types;
* which service areas may deserve greater improvement priority; and
* whether higher service ratings are generally associated with higher customer advocacy.

The **Overall** view provides a company-wide summary, while the segment views support more targeted recommendations.

### How to Interact

* Use the dropdown menu to switch between **Overall, Government, NPO and Private** results.
* Hover over a bar to view the service dimension, exact Spearman correlation and interpretation band.
* Compare the length of the bars to identify the strongest and weakest associations.
* Use the fixed scale from −1 to +1 to compare the correlation strength fairly across client types.

> **Interpretation note:** Correlation shows association, not causation. A strong correlation does not prove that improving a service dimension will directly cause NPS Rating to increase. Other factors not included in the dataset may also influence customer loyalty.



In [16]:
from plotly.subplots import make_subplots
import plotly.graph_objects as go
import numpy as np


def kay_fig3(d, segment):

    # Apply selected client segment
    if segment == "Overall":
        g = d.copy()
    else:
        g = d[d["TYPE"].eq(segment)].copy()

    if g.empty:
        return go.Figure().update_layout(
            title="No data match the selected filters",
            template="plotly_white",
            height=780
        )

    service_pairs = [
        (
            "PRESALES AND PARTNERSHIP",
            "Presales & Partnership"
        ),
        (
            "TECHNICAL EXPERTISE",
            "Technical Expertise"
        ),
        (
            "PROJECT DELIVERY",
            "Project Delivery"
        ),
        (
            "POST-SALES SUPPORT",
            "Post-Sales Support"
        )
    ]

    type_colours = {
        "Govt": "#636EFA",
        "NPO": "#00CC96",
        "Private": "#EF553B"
    }

    # Calculate correlations before creating subplot titles
    correlations = {}

    for column, label in service_pairs:
        valid = g[[column, "NPS RATING"]].dropna()

        correlations[label] = valid[column].corr(
            valid["NPS RATING"],
            method="spearman"
        )

    fig = make_subplots(
        rows=2,
        cols=2,
        subplot_titles=[
            label for _, label in service_pairs
        ],
        horizontal_spacing=0.12,
        vertical_spacing=0.18
    )

    # Fixed random seed keeps jitter unchanged
    rng = np.random.default_rng(42)

    for position, (column, label) in enumerate(
        service_pairs,
        start=1
    ):

        row = 1 if position <= 2 else 2
        col = 1 if position % 2 == 1 else 2

        plot_data = g[
            [column, "NPS RATING", "TYPE", "YEAR"]
        ].dropna().copy()

        # Jitter only changes point positions visually
        plot_data["Rating Jitter"] = (
            plot_data[column]
            + rng.uniform(-0.12, 0.12, len(plot_data))
        )

        plot_data["NPS Jitter"] = (
            plot_data["NPS RATING"]
            + rng.uniform(-0.14, 0.14, len(plot_data))
        )

        # Add client observations
        for client_type in sorted(
            plot_data["TYPE"].unique()
        ):
            client_data = plot_data[
                plot_data["TYPE"].eq(client_type)
            ]

            fig.add_trace(
                go.Scattergl(
                    x=client_data["Rating Jitter"],
                    y=client_data["NPS Jitter"],
                    mode="markers",
                    name=client_type,
                    legendgroup=client_type,
                    showlegend=(position == 1),
                    marker={
                        "size": 8,
                        "opacity": 0.55,
                        "color": type_colours.get(
                            client_type,
                            "#7A869A"
                        ),
                        "line": {
                            "width": 0.5,
                            "color": "white"
                        }
                    },
                    customdata=client_data[
                        [column, "NPS RATING", "TYPE", "YEAR"]
                    ].to_numpy(),
                    hovertemplate=(
                        f"<b>{label}</b><br>"
                        "Client Type: %{customdata[2]}<br>"
                        "Year: %{customdata[3]}<br>"
                        "Service Rating: %{customdata[0]:.0f}/5<br>"
                        "NPS Rating: %{customdata[1]:.0f}/10"
                        "<extra></extra>"
                    )
                ),
                row=row,
                col=col
            )

        # Calculate average NPS at each rating level
        rating_means = (
            plot_data
            .groupby(column, as_index=False)["NPS RATING"]
            .mean()
            .sort_values(column)
        )

        # Add average trend line
        fig.add_trace(
            go.Scatter(
                x=rating_means[column],
                y=rating_means["NPS RATING"],
                mode="lines+markers",
                name="Mean NPS by Rating",
                legendgroup="Mean NPS",
                showlegend=(position == 1),
                line={
                    "color": "#172B4D",
                    "width": 3
                },
                marker={
                    "size": 8,
                    "color": "#172B4D"
                },
                hovertemplate=(
                    "Service Rating: %{x:.0f}/5<br>"
                    "Average NPS: %{y:.2f}/10"
                    "<extra>Mean NPS</extra>"
                )
            ),
            row=row,
            col=col
        )

        # Add small correlation label
        axis_suffix = "" if position == 1 else str(position)

        correlation = correlations[label]

        correlation_text = (
            "Spearman ρ: unavailable"
            if np.isnan(correlation)
            else f"Spearman ρ = {correlation:+.2f}"
        )

        fig.add_annotation(
            x=0.97,
            y=0.08,
            xref=f"x{axis_suffix} domain",
            yref=f"y{axis_suffix} domain",
            text=correlation_text,
            showarrow=False,
            xanchor="right",
            yanchor="bottom",
            font={
                "size": 12,
                "color": "#172B4D"
            },
            bgcolor="rgba(255,255,255,0.85)",
            bordercolor="#D9E2EC",
            borderwidth=1,
            borderpad=5
        )

        fig.update_xaxes(
            title_text="Service Rating",
            tickmode="array",
            tickvals=[1, 2, 3, 4, 5],
            range=[0.65, 5.35],
            row=row,
            col=col
        )

        fig.update_yaxes(
            title_text="NPS Rating",
            tickmode="array",
            tickvals=[0, 2, 4, 6, 8, 10],
            range=[-0.5, 10.5],
            row=row,
            col=col
        )

    fig.update_layout(
        title={
            "text": (
                "Service Ratings Against NPS"
                f" — {segment}"
                "<br><sup>"
                "Points show client-year responses; "
                "dark lines show average NPS at each rating"
                "</sup>"
            ),
            "x": 0.5,
            "xanchor": "center"
        },
        template="plotly_white",
        height=780,
        hovermode="closest",
        legend={
            "title": "Client Type",
            "orientation": "h",
            "yanchor": "bottom",
            "y": 1.08,
            "xanchor": "center",
            "x": 0.5
        },
        margin={
            "t": 145,
            "l": 70,
            "r": 40,
            "b": 70
        }
    )

    return fig

In [17]:
fig3 = kay_fig3(
    df,
    "Overall"
)

fig3.show()

### Key Findings from Visualisation 3

#### Overall Relationships

All four service dimensions have a **positive correlation** with NPS Rating. This means that clients who provide higher service ratings also generally provide higher NPS ratings.

The strongest overall relationships are:

* **Technical Expertise:** 0.544
* **Project Delivery:** 0.538
* **Post-Sales Support:** 0.444
* **Presales & Partnership:** 0.412

Technical Expertise and Project Delivery have very similar results, indicating that both are important to customer advocacy. Clients may be more willing to recommend the company when they trust its technical capabilities and are satisfied with how their projects are completed.

#### Findings by Client Type

| Client Type    | Strongest Associations                                 | Interpretation                                                                                                     |
| -------------- | ------------------------------------------------------ | ------------------------------------------------------------------------------------------------------------------ |
| **Government** | Project Delivery: 0.559; Technical Expertise: 0.522    | Successful project execution and technical capability are most closely associated with Government client advocacy. |
| **NPO**        | Project Delivery: 0.670; Presales & Partnership: 0.619 | NPO loyalty is strongly associated with both project execution and the relationship established before delivery.   |
| **Private**    | Technical Expertise: 0.571; Project Delivery: 0.521    | Private clients appear to place greater importance on technical capability and reliable project completion.        |

For Government clients, Project Delivery has the strongest relationship with NPS Rating at **0.559**, followed by Technical Expertise at **0.522**. Presales & Partnership has the weakest Government relationship at **0.246**.

For NPO clients, Project Delivery records the strongest correlation in the analysis at **0.670**. Presales & Partnership is also strongly associated with NPS Rating at **0.619**. Technical Expertise and Post-Sales Support have weaker relationships at **0.369** and **0.281** respectively.

For Private clients, Technical Expertise has the strongest relationship at **0.571**, followed by Project Delivery at **0.521**. Post-Sales Support and Presales & Partnership show positive but weaker relationships at **0.464** and **0.426**.

#### Combined Interpretation with Visualisation 1

Correlation strength should be considered together with the average ratings from Visualisation 1.

* **Government Project Delivery** has a strong correlation with NPS and a high average rating of 4.35. It should therefore be treated as a strength to maintain.
* **NPO Presales & Partnership** has a relatively low average rating of 3.76 but a strong correlation of 0.619. This makes it an important improvement opportunity.
* **Private Technical Expertise** has the strongest association with Private NPS. Its quality should be maintained and strengthened.
* **Private Post-Sales Support** has the segment’s lowest average rating and a positive correlation of 0.464, making it another practical improvement area.

#### Business Implications

The company should prioritise different service areas for different client types:

1. Maintain strong Project Delivery and Technical Expertise across all segments.
2. Improve Presales & Partnership for NPO clients.
3. Protect Project Delivery quality for Government clients.
4. Strengthen Technical Expertise and Post-Sales Support for Private clients.
5. Use both average ratings and correlation strength when deciding where improvements are most valuable.

#### Limitations

Correlation measures association and does not prove causation. The results do not confirm that improving one service dimension will directly increase NPS Rating.

The client segments also have different numbers of observations, so correlations from smaller groups may be less stable. In addition, some clients appear across multiple years, meaning the observations are not completely independent.

The results should therefore be treated as evidence of possible satisfaction drivers rather than proof of direct cause and effect.



## Interactive satisfaction dashboard

The dashboard combines the three visualisations into one decision view:

1. the grouped bar chart compares **service performance**;
2. the heatmap compares **loyalty levels over time**; and
3. the correlation chart highlights potential **loyalty drivers**.

The KPI cards provide an overall snapshot, while the chart controls allow users to explore periods and client types. Run the cell below to launch the dashboard locally.

In [18]:
import numpy as np
import pandas as pd
import plotly.express as px
import plotly.graph_objects as go

from dash import Dash, Input, Output, dcc, html


# -------------------------------------------------
# 1. Dashboard data and labels
# -------------------------------------------------

dashboard_df = df.copy()
dashboard_df["YEAR"] = pd.to_numeric(
    dashboard_df["YEAR"], errors="coerce"
).astype("Int64")

service_columns = [
    "PRESALES AND PARTNERSHIP",
    "TECHNICAL EXPERTISE",
    "PROJECT DELIVERY",
    "POST-SALES SUPPORT"
]

service_label_map = {
    "PRESALES AND PARTNERSHIP": "Presales & Partnership",
    "TECHNICAL EXPERTISE": "Technical Expertise",
    "PROJECT DELIVERY": "Project Delivery",
    "POST-SALES SUPPORT": "Post-Sales Support"
}

service_colour_map = {
    "Presales & Partnership": "#636EFA",
    "Technical Expertise": "#EF553B",
    "Project Delivery": "#00CC96",
    "Post-Sales Support": "#AB63FA"
}

client_types = sorted(dashboard_df["TYPE"].dropna().unique())
years = sorted(dashboard_df["YEAR"].dropna().astype(int).unique())


# -------------------------------------------------
# 2. Reusable styles
# -------------------------------------------------

page_style = {
    "backgroundColor": "#F4F6FA",
    "minHeight": "100vh",
    "padding": "24px",
    "fontFamily": "Segoe UI, Arial, sans-serif",
    "color": "#172B4D"
}

content_style = {
    "maxWidth": "1500px",
    "margin": "0 auto"
}

navbar_style = {
    "position": "sticky",
    "top": "0",
    "zIndex": "1000",
    "display": "grid",
    "gridTemplateColumns": "repeat(auto-fit, minmax(190px, 1fr))",
    "gap": "14px",
    "alignItems": "end",
    "backgroundColor": "#172B4D",
    "padding": "18px",
    "borderRadius": "12px",
    "boxShadow": "0 3px 12px rgba(0, 0, 0, 0.15)",
    "marginBottom": "18px"
}

filter_label_style = {
    "display": "block",
    "color": "white",
    "fontSize": "13px",
    "fontWeight": "600",
    "marginBottom": "6px"
}

metric_grid_style = {
    "display": "grid",
    "gridTemplateColumns": "repeat(auto-fit, minmax(210px, 1fr))",
    "gap": "16px",
    "marginBottom": "20px"
}

metric_card_style = {
    "backgroundColor": "white",
    "padding": "18px",
    "borderRadius": "12px",
    "boxShadow": "0 2px 8px rgba(0, 0, 0, 0.08)"
}

chart_card_style = {
    "backgroundColor": "white",
    "padding": "12px",
    "borderRadius": "12px",
    "boxShadow": "0 2px 8px rgba(0, 0, 0, 0.08)"
}

bottom_grid_style = {
    "display": "grid",
    "gridTemplateColumns": "repeat(auto-fit, minmax(480px, 1fr))",
    "gap": "20px",
    "marginTop": "20px"
}

graph_config = {
    "displaylogo": False,
    "responsive": True,
    "scrollZoom": True,
    "toImageButtonOptions": {
        "format": "png",
        "filename": "client_satisfaction_dashboard",
        "scale": 2
    }
}


# -------------------------------------------------
# 3. Figure-building functions
# -------------------------------------------------

def make_empty_figure(message):
    empty_figure = go.Figure()
    empty_figure.add_annotation(
        text=message,
        x=0.5,
        y=0.5,
        xref="paper",
        yref="paper",
        showarrow=False,
        font={"size": 16, "color": "#5E6C84"}
    )
    empty_figure.update_layout(
        template="plotly_white",
        xaxis={"visible": False},
        yaxis={"visible": False},
        height=420
    )
    return empty_figure


def make_service_figure(filtered_df, selected_services, filter_text):
    service_summary = (
        filtered_df
        .groupby("TYPE", observed=True)[selected_services]
        .mean()
        .reset_index()
    )

    record_counts = filtered_df.groupby("TYPE", observed=True).size()

    service_long = service_summary.melt(
        id_vars="TYPE",
        value_vars=selected_services,
        var_name="Service Column",
        value_name="Average Rating"
    )

    service_long["Service Dimension"] = service_long[
        "Service Column"
    ].map(service_label_map)

    service_long["Records"] = service_long["TYPE"].map(record_counts)

    figure = px.bar(
        service_long,
        x="TYPE",
        y="Average Rating",
        color="Service Dimension",
        barmode="group",
        text="Average Rating",
        custom_data=["Records"],
        color_discrete_map=service_colour_map,
        category_orders={
            "TYPE": client_types,
            "Service Dimension": [
                service_label_map[column]
                for column in selected_services
            ]
        },
        title=f"Service Ratings — {filter_text}"
    )

    figure.update_traces(
        texttemplate="%{y:.2f}",
        textposition="outside",
        cliponaxis=False,
        hovertemplate=(
            "Client Type: %{x}<br>"
            "Average Rating: %{y:.2f}<br>"
            "Records: %{customdata[0]}"
            "<extra>%{fullData.name}</extra>"
        )
    )

    figure.update_layout(
        template="plotly_white",
        height=560,
        yaxis={"range": [0, 5], "title": "Average Rating"},
        xaxis_title="Client Type",
        legend_title="Service Dimension",
        title_x=0.5,
        margin={"t": 80, "r": 30, "b": 60, "l": 60}
    )

    return figure


def make_nps_figure(filtered_df, selected_year, filter_text):
    nps_summary = (
        filtered_df
        .groupby(["TYPE", "YEAR"], observed=True)["NPS RATING"]
        .agg(["mean", "count"])
        .reset_index()
    )

    value_matrix = nps_summary.pivot(
        index="TYPE",
        columns="YEAR",
        values="mean"
    )

    count_matrix = nps_summary.pivot(
        index="TYPE",
        columns="YEAR",
        values="count"
    )

    displayed_types = [
        client_type
        for client_type in client_types
        if client_type in value_matrix.index
    ]

    if selected_year == "All":
        displayed_periods = [
            year for year in years if year in value_matrix.columns
        ]

        overall_nps = filtered_df.groupby(
            "TYPE", observed=True
        )["NPS RATING"].mean()

        overall_count = filtered_df.groupby(
            "TYPE", observed=True
        ).size()

        value_matrix["Overall"] = overall_nps
        count_matrix["Overall"] = overall_count
        displayed_periods = displayed_periods + ["Overall"]
    else:
        displayed_periods = [int(selected_year)]

    value_matrix = value_matrix.reindex(
        index=displayed_types,
        columns=displayed_periods
    )

    count_matrix = count_matrix.reindex(
        index=displayed_types,
        columns=displayed_periods
    )

    heatmap_text = np.array([
        ["" if pd.isna(value) else f"{value:.2f}" for value in row]
        for row in value_matrix.to_numpy()
    ])

    figure = go.Figure(
        data=go.Heatmap(
            z=value_matrix.to_numpy(),
            x=[str(period) for period in displayed_periods],
            y=displayed_types,
            text=heatmap_text,
            texttemplate="%{text}",
            textfont={"size": 16},
            customdata=count_matrix.to_numpy(),
            colorscale=[
                [0.00, "#E76F51"],
                [0.50, "#F4D35E"],
                [1.00, "#2A9D8F"]
            ],
            zmin=7,
            zmax=9,
            colorbar={"title": "Average<br>NPS Rating"},
            hovertemplate=(
                "Client Type: %{y}<br>"
                "Period: %{x}<br>"
                "Average NPS Rating: %{z:.2f}<br>"
                "Records: %{customdata:.0f}"
                "<extra></extra>"
            )
        )
    )

    figure.update_layout(
        template="plotly_white",
        title=f"Average NPS Rating — {filter_text}",
        title_x=0.5,
        xaxis_title="Period",
        yaxis_title="Client Type",
        height=470,
        margin={"t": 80, "r": 30, "b": 60, "l": 70}
    )

    return figure


def make_driver_figure(filtered_df, selected_services, filter_text):
    correlation_rows = []

    for column in selected_services:
        pair = filtered_df[[column, "NPS RATING"]].dropna()

        if len(pair) >= 3:
            correlation = pair[column].rank().corr(
                pair["NPS RATING"].rank()
            )
        else:
            correlation = np.nan

        correlation_rows.append({
            "Service Dimension": service_label_map[column],
            "Correlation": correlation
        })

    correlation_df = (
        pd.DataFrame(correlation_rows)
        .dropna(subset=["Correlation"])
        .sort_values("Correlation", ascending=False)
    )

    if correlation_df.empty:
        return make_empty_figure(
            "Not enough records to calculate correlations."
        )

    figure = go.Figure(
        go.Bar(
            x=correlation_df["Correlation"],
            y=correlation_df["Service Dimension"],
            orientation="h",
            marker_color="#2A9D8F",
            text=correlation_df["Correlation"].map(
                lambda value: f"{value:+.2f}"
            ),
            textposition="outside",
            customdata=np.repeat(len(filtered_df), len(correlation_df)),
            hovertemplate=(
                "Service Dimension: %{y}<br>"
                "Spearman Correlation: %{x:.3f}<br>"
                "Filtered Records: %{customdata}"
                "<extra></extra>"
            )
        )
    )

    figure.add_vrect(
        x0=0,
        x1=0.3,
        fillcolor="#F6EFD8",
        opacity=0.35,
        line_width=0,
        layer="below"
    )

    figure.add_vrect(
        x0=0.3,
        x1=0.5,
        fillcolor="#E7F2EF",
        opacity=0.45,
        line_width=0,
        layer="below"
    )

    figure.add_vrect(
        x0=0.5,
        x1=1,
        fillcolor="#D5ECE7",
        opacity=0.55,
        line_width=0,
        layer="below"
    )

    figure.update_layout(
        template="plotly_white",
        title=f"Service Dimensions Associated with NPS — {filter_text}",
        title_x=0.5,
        xaxis={
            "title": "Spearman Correlation with NPS Rating",
            "range": [-1, 1],
            "zeroline": True,
            "zerolinewidth": 2,
            "zerolinecolor": "#5E6C84"
        },
        yaxis={
            "title": "Service Dimension",
            "autorange": "reversed"
        },
        height=470,
        margin={"t": 90, "r": 60, "b": 60, "l": 150}
    )

    return figure


# -------------------------------------------------
# 4. Build the Dash application
# -------------------------------------------------

app = Dash(__name__)
app.title = "Client Satisfaction Dashboard"

app.layout = html.Div(
    style=page_style,
    children=[
        html.Div(
            style=content_style,
            children=[
                html.Nav(
                    style=navbar_style,
                    children=[
                        html.Div(
                            children=[
                                html.H2(
                                    "Client Satisfaction Dashboard",
                                    style={
                                        "color": "white",
                                        "margin": "0 0 4px 0"
                                    }
                                ),
                                html.P(
                                    id="filter-summary",
                                    style={
                                        "color": "#D6E0F0",
                                        "margin": "0",
                                        "fontSize": "13px"
                                    }
                                )
                            ]
                        ),

                        html.Div(
                            children=[
                                html.Label(
                                    "Client Type",
                                    style=filter_label_style
                                ),
                                dcc.Dropdown(
                                    id="client-type-filter",
                                    options=(
                                        [{"label": "All Client Types", "value": "All"}]
                                        + [
                                            {"label": value, "value": value}
                                            for value in client_types
                                        ]
                                    ),
                                    value="All",
                                    clearable=False
                                )
                            ]
                        ),

                        html.Div(
                            children=[
                                html.Label(
                                    "Year",
                                    style=filter_label_style
                                ),
                                dcc.Dropdown(
                                    id="year-filter",
                                    options=(
                                        [{"label": "All Years", "value": "All"}]
                                        + [
                                            {"label": str(year), "value": year}
                                            for year in years
                                        ]
                                    ),
                                    value="All",
                                    clearable=False
                                )
                            ]
                        ),

                        html.Div(
                            children=[
                                html.Label(
                                    "Service Dimensions",
                                    style=filter_label_style
                                ),
                                dcc.Dropdown(
                                    id="service-filter",
                                    options=[
                                        {
                                            "label": service_label_map[column],
                                            "value": column
                                        }
                                        for column in service_columns
                                    ],
                                    value=service_columns,
                                    multi=True,
                                    clearable=False
                                )
                            ]
                        ),

                        html.Button(
                            "Reset Filters",
                            id="reset-filters",
                            n_clicks=0,
                            style={
                                "height": "38px",
                                "padding": "0 18px",
                                "border": "1px solid #D6E0F0",
                                "borderRadius": "6px",
                                "backgroundColor": "white",
                                "color": "#172B4D",
                                "fontWeight": "600",
                                "cursor": "pointer"
                            }
                        )
                    ]
                ),

                html.Div(
                    style=metric_grid_style,
                    children=[
                        html.Div(
                            style=metric_card_style,
                            children=[
                                html.P(
                                    "Average Service Rating",
                                    style={"color": "#5E6C84", "margin": "0"}
                                ),
                                html.H2(
                                    id="overall-service-value",
                                    style={"marginBottom": "0"}
                                )
                            ]
                        ),

                        html.Div(
                            style=metric_card_style,
                            children=[
                                html.P(
                                    "Average NPS Rating",
                                    style={"color": "#5E6C84", "margin": "0"}
                                ),
                                html.H2(
                                    id="average-nps-value",
                                    style={"marginBottom": "0"}
                                )
                            ]
                        ),

                        html.Div(
                            style=metric_card_style,
                            children=[
                                html.P(
                                    "Strongest Selected Dimension",
                                    style={"color": "#5E6C84", "margin": "0"}
                                ),
                                html.H3(
                                    id="strongest-service-value",
                                    style={"marginBottom": "0"}
                                )
                            ]
                        ),

                        html.Div(
                            style=metric_card_style,
                            children=[
                                html.P(
                                    "Filtered Records",
                                    style={"color": "#5E6C84", "margin": "0"}
                                ),
                                html.H2(
                                    id="record-count-value",
                                    style={"marginBottom": "0"}
                                )
                            ]
                        )
                    ]
                ),

                html.Div(
                    style=chart_card_style,
                    children=[
                        dcc.Loading(
                            dcc.Graph(
                                id="service-rating-chart",
                                config=graph_config,
                                responsive=True
                            )
                        )
                    ]
                ),

                html.Div(
                    style=bottom_grid_style,
                    children=[
                        html.Div(
                            style=chart_card_style,
                            children=[
                                dcc.Loading(
                                    dcc.Graph(
                                        id="nps-heatmap",
                                        config=graph_config,
                                        responsive=True
                                    )
                                )
                            ]
                        ),

                        html.Div(
                            style=chart_card_style,
                            children=[
                                dcc.Loading(
                                    dcc.Graph(
                                        id="nps-driver-chart",
                                        config=graph_config,
                                        responsive=True
                                    )
                                )
                            ]
                        )
                    ]
                ),

                html.P(
                    (
                        "Correlations indicate association and do not "
                        "necessarily establish causation."
                    ),
                    style={
                        "textAlign": "center",
                        "color": "#7A869A",
                        "fontSize": "12px",
                        "marginTop": "18px"
                    }
                )
            ]
        )
    ]
)


# -------------------------------------------------
# 5. Reset the navbar filters
# -------------------------------------------------

@app.callback(
    Output("client-type-filter", "value"),
    Output("year-filter", "value"),
    Output("service-filter", "value"),
    Input("reset-filters", "n_clicks"),
    prevent_initial_call=True
)
def reset_filters(_):
    return "All", "All", service_columns


# -------------------------------------------------
# 6. Update the complete dashboard
# -------------------------------------------------

@app.callback(
    Output("service-rating-chart", "figure"),
    Output("nps-heatmap", "figure"),
    Output("nps-driver-chart", "figure"),
    Output("overall-service-value", "children"),
    Output("average-nps-value", "children"),
    Output("strongest-service-value", "children"),
    Output("record-count-value", "children"),
    Output("filter-summary", "children"),
    Input("client-type-filter", "value"),
    Input("year-filter", "value"),
    Input("service-filter", "value")
)
def update_dashboard(selected_client_type, selected_year, selected_services):
    selected_services = [
        column
        for column in (selected_services or service_columns)
        if column in service_columns
    ]

    filtered_df = dashboard_df.copy()

    if selected_client_type != "All":
        filtered_df = filtered_df[
            filtered_df["TYPE"] == selected_client_type
        ]

    if selected_year != "All":
        filtered_df = filtered_df[
            filtered_df["YEAR"] == int(selected_year)
        ]

    client_text = (
        "All Client Types"
        if selected_client_type == "All"
        else selected_client_type
    )

    year_text = (
        "All Years"
        if selected_year == "All"
        else str(selected_year)
    )

    filter_text = f"{client_text}, {year_text}"

    if filtered_df.empty:
        empty = make_empty_figure("No records match the selected filters.")
        return (
            empty,
            empty,
            empty,
            "No data",
            "No data",
            "No data",
            "0",
            f"Showing: {filter_text}"
        )

    service_figure = make_service_figure(
        filtered_df,
        selected_services,
        filter_text
    )

    nps_figure = make_nps_figure(
        filtered_df,
        selected_year,
        filter_text
    )

    driver_figure = kay_fig3(
        filtered_df,
        "Overall"
    )

    driver_figure.update_layout(
        title={
            "text": (
                "Service Ratings Against NPS"
                f" — {filter_text}"
                "<br><sup>"
                "Points show client-year responses; "
                "dark lines show average NPS at each rating"
                "</sup>"
            ),
            "x": 0.5,
            "xanchor": "center"
        }
    )

    average_service = (
        filtered_df[selected_services]
        .mean(axis=1)
        .mean()
    )

    average_nps = filtered_df["NPS RATING"].mean()
    service_means = filtered_df[selected_services].mean()
    strongest_column = service_means.idxmax()
    strongest_name = service_label_map[strongest_column]
    strongest_rating = service_means.loc[strongest_column]

    return (
        service_figure,
        nps_figure,
        driver_figure,
        f"{average_service:.2f} / 5",
        f"{average_nps:.2f} / 10",
        f"{strongest_name} ({strongest_rating:.2f} / 5)",
        f"{len(filtered_df):,}",
        f"Showing: {filter_text}"
    )


# -------------------------------------------------
# 7. Run the dashboard
# -------------------------------------------------

app.run(debug=True)

## Conclusion and Strategic Recommendations

### Overall Analytical Conclusion

The three visualisations suggest that customer satisfaction is generally positive, but the factors associated with loyalty differ across client segments. Therefore, the company should not simply focus on the lowest rating or apply the same improvement strategy to every client.

A more useful approach is to consider two factors together:

1. **Performance:** Whether the service dimension currently has a high or low average rating.
2. **Importance:** How strongly that dimension is associated with NPS Rating.

A service dimension with a **low rating and strong NPS association** represents a more urgent improvement opportunity. A dimension with a **high rating and strong association** represents a strength that should be protected.

### Government Clients: Protect a Strong Delivery Experience

Government clients report the highest overall ratings for Presales & Partnership, Technical Expertise and Project Delivery. Their average NPS Rating is also strong at **8.11**, reaching **8.65 in 2025**.

Project Delivery is both:

* the highest-rated Government service dimension at **4.35**; and
* the dimension most strongly associated with Government NPS Rating at **0.559**.

This combination suggests that Project Delivery is not simply performing well-it is also closely connected to how Government clients evaluate their overall relationship with the company.

Therefore, Project Delivery should be treated as a strength to protect rather than an area requiring major corrective action. The company should investigate what contributed to the Government Project Delivery rating increasing from **3.67 in 2021 to 4.75 in 2025**, and determine whether these practices can be standardised.

Government Post-Sales Support is lower at **3.91**, but its NPS association is also weaker than Project Delivery. It remains an improvement opportunity, although it may not have the same strategic importance as maintaining delivery quality.

### NPO Clients: Strong Outcomes but a Weak Initial Experience

NPO clients have the highest overall average NPS Rating at **8.12**, although this is only 0.01 above Government clients. They also report strong Project Delivery and Post-Sales Support ratings of **4.26** and **4.12**.

However, Presales & Partnership is their weakest service dimension at **3.76** and has a strong correlation of **0.619** with NPS Rating.

This creates an important finding: NPO clients are currently positive about the company overall, but the initial relationship-building stage may be a hidden weakness. Because Presales & Partnership is strongly associated with their NPS Rating, continued weakness in this area could prevent the company from building stronger advocacy among NPO clients.

NPO Presales & Partnership should therefore be treated as a high-priority improvement area. Actions could include clearer expectation setting, more regular early-stage communication and greater involvement of NPO clients during solution planning.

NPO ratings also fluctuate more noticeably between years. This may indicate an inconsistent experience, although the result could also be influenced by a smaller number of records. The company should examine individual NPO responses before making large operational changes.

### Private Clients: Loyalty Gap May Extend Beyond Service Ratings

Private clients have the lowest overall average NPS Rating at **7.81**, but their service ratings are not dramatically lower than those of the other segments. Their Project Delivery rating, for example, is still positive at **4.01**.

This creates a possible gap: Private clients appear reasonably satisfied with the measured service dimensions, but they are less willing to recommend the company.

Technical Expertise has the strongest association with Private NPS Rating at **0.571**, while Post-Sales Support has the lowest average rating at **3.80** and a positive association of **0.464**.

The company should therefore use a two-part strategy for Private clients:

1. Maintain and strengthen Technical Expertise because it is most closely associated with their advocacy.
2. Improve Post-Sales Support because it is their weakest reported service experience.

However, the lower Private NPS Rating may not be fully explained by these four service dimensions. Other factors not included in the dataset—such as price, perceived value, product suitability or competitor alternatives—may also influence their willingness to recommend the company. Further investigation is needed before assuming that service improvements alone will remove the loyalty gap.

### Company-Wide Pattern in 2025

All three client types record their highest NPS Rating in 2025. Project Delivery also reaches its highest value for every client type during the same year.

Because the improvement occurs across multiple segments, it may reflect a company-wide change rather than a segment-specific event. Possible explanations could include improved project processes, staff capability or changes in the clients surveyed.

The chart cannot confirm which explanation is correct. The company should compare its 2025 operational practices and client composition with previous years before treating the increase as a lasting improvement.

### Strategic Priorities

Based on both average performance and NPS association, the recommended priorities are:

| Priority                                 | Evidence                                                                      | Recommended Response                                                           |
| ---------------------------------------- | ----------------------------------------------------------------------------- | ------------------------------------------------------------------------------ |
| **Protect Project Delivery**             | Highest-rated dimension across every segment and strongly associated with NPS | Maintain quality and investigate successful 2025 practices                     |
| **Improve NPO Presales & Partnership**   | Low rating of 3.76 and strong NPS association of 0.619                        | Strengthen early communication and relationship-building                       |
| **Improve Private Post-Sales Support**   | Lowest Private service rating at 3.80                                         | Improve follow-up, support access and issue resolution                         |
| **Maintain Private Technical Expertise** | Strongest Private NPS association at 0.571                                    | Protect technical quality and communicate technical value clearly              |
| **Investigate the 2025 improvement**     | Highest NPS and Project Delivery ratings across all segments                  | Identify whether operational changes or sample composition caused the increase |

### Connection to the Main Problem

This analysis identifies which client segments maintain stronger customer satisfaction, but it cannot independently determine which segments should receive the highest business priority.

The satisfaction findings should be combined with the team’s profitability results using the following framework:

* **High profit and high satisfaction:** Protect and grow the relationship.
* **High profit and low satisfaction:** Prioritise service recovery because valuable clients may be at risk.
* **Low profit and high satisfaction:** Explore cross-selling, pricing or cost optimisation opportunities.
* **Low profit and low satisfaction:** Reassess the commercial value of continued investment.

Government and NPO clients currently show stronger overall satisfaction and advocacy, while Private clients present clearer improvement opportunities. However, the final recommendation depends on whether these segments also generate sufficient gross profit and gross margin.

### Final Limitation

These findings identify patterns and associations rather than proven causes. Average ratings may hide differences between individual clients, the client segments have unequal sample sizes, and some clients appear across multiple years.

The analysis also uses average NPS Rating rather than the formal Net Promoter Score and does not contain actual renewal or churn outcomes. Therefore, NPS should be interpreted as an indicator of customer advocacy—not proof of future retention.

The results provide evidence for prioritisation, but the company should validate the findings using client-level feedback, sample sizes, profitability measures and actual retention outcomes before making major strategic decisions.
